Fix cert expiration issue for aggregator client

praveenkumar opened the pull request:
During 4.17.0-ec.1 testing we found out that it might be possible that
kubelet certs are not expired but aggregator one is and due to that our
apiserver not able to get the node info and fails with following error.

```
INFO Verifying validity of the kubelet certificates...
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-client-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2025-07-05T03:48:39+00:00
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-server-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2025-07-05T03:49:24+00:00
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2024-07-11T05:50:37+00:00
DEBU Certs have expired, they were valid till: 11 Jul 24 05:50 +0000
```

This PR make sure we also wait to recover the api server related certs
before checking the apiserver is responding.

praveenkumar committed:
cluster: Update the location of aggregator-client certificate

aggregator-client-ca is a CA bundle, which contains a single signer
- aggregator-client-signer. The target certificate is created from
this signer - aggregator-client which can be tracked for cert expire
purpose.

/etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt
represent the CA bundle which contains the signer
aggregator-client-signer and create the target cert for
aggregator-client which is synced to /etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/aggregator-client/tls.crt

During testing of 4.17.0.ec.1 found out that
`secrets/aggregator-client/tls.crt` is synced before `configmaps/aggregator-client-ca/ca-bundle.crt` which is due to the certsyncer controller.

If we wait for `configmaps/aggregator-client-ca/ca-bundle.crt` then it
takes more time ~10 mins but `secrets/aggregator-client/tls.crt` appear
early and good signal about cert is rotated. That's the reason we are
changing the location of it in this PR.

```
$ ls -l /etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/
total 0
drwxr-xr-x. 2 root root  36 Jul 18 10:43 aggregator-client
$ ls -l /etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/
total 0
drwxr-xr-x. 2 root root 27 Jul 18 10:44 aggregator-client-ca
```

praveenkumar committed:
Add WaitForKubeAPIServerCertRenewal helper

During 4.17.0-ec.1 testing we found out that it might be possible that
kubelet certs are not expired but aggreator one is and due to that our
apiserver not able to get the node info and fails with following error.

```
INFO Verifying validity of the kubelet certificates...
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-client-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2025-07-05T03:48:39+00:00
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /var/lib/kubelet/pki/kubelet-server-current.pem -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2025-07-05T03:49:24+00:00
DEBU Running SSH command: date --date="$(sudo openssl x509 -in /etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt -noout -enddate | cut -d= -f 2)" --iso-8601=seconds
DEBU SSH command results: err: <nil>, output: 2024-07-11T05:50:37+00:00
DEBU Certs have expired, they were valid till: 11 Jul 24 05:50 +0000

DEBU Running SSH command: timeout 5s oc get nodes --context admin --cluster crc --kubeconfig /opt/kubeconfig
DEBU SSH command results: err: Process exited with status 1, output:
DEBU E0722 10:21:40.601631   10967 memcache.go:265] couldn't get current server API group list: Get "https://api.crc.testing:6443/api?timeout=32s": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-07-22T10:21:40Z is after 2024-07-10T05:27:06Z
E0722 10:21:40.604575   10967 memcache.go:265] couldn't get current server API group list: Get "https://api.crc.testing:6443/api?timeout=32s": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-07-22T10:21:40Z is after 2024-07-10T05:27:06Z
```

This PR make sure we also wait to recover the api server related certs
before checking the apiserver is responding.

cfergeau commented on pkg/crc/cluster/cluster.go:
```diff
@@ -32,7 +32,7 @@ const (
 	KubeletClientCert = "/var/lib/kubelet/pki/kubelet-client-current.pem"
 
-	AggregatorClientCert = "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt"
+	AggregatorClientCert = "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/aggregator-client/tls.crt"
```
The commit log says
> During testing of 4.17.0.ec.1 found out that `secrets/aggregator-client/tls.crt` is regenerated as part of auto cert rotation before `configmaps/aggregator-client-ca/ca-bundle.crt`.

Is `configmaps/aggregator-client-ca` really regenerated after `secrets/aggregator-client`? Or does this happen in the opposite order?

cfergeau replied:
As a preparatory commit, I'd remove https://github.com/crc-org/crc/blob/2fa7958844be56353ebfd6c104281a77a5a24b5c/pkg/crc/machine/start.go#L605-L626 as this change has an impact on this code.

praveenkumar replied:
Yes it is really regenerated after `secrets/aggregator-client`, I ssh'ed to the node and checked this.

cfergeau replied:
It's quite odd, as it would make more sense if `secrets/aggregator-client` used the new CA, not an older one.
If `configmaps/aggregator-client-ca` is the last to be regenerated, why do we change to check for `secrets/aggregator-client`? When `aggregator-client-ca` is fully renewed, `aggregator-client` should also be fine?

praveenkumar replied:
> As a preparatory commit, I'd remove 

Removed as first commit

>It's quite odd, as it would make more sense if secrets/aggregator-client used the new CA, not an older one.
If configmaps/aggregator-client-ca is the last to be regenerated, why do we change to check for secrets/aggregator-client? When aggregator-client-ca is fully renewed, aggregator-client should also be fine?

During that internal conversation figure out the due to certsyncer controller it might happen that cert files sync on the disk is out of order, added more context in the commit log.

cfergeau replied:
>  I ssh'ed to the node and checked this.

With `ls -al`, or with something like `$ cat /etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/aggregator-client/tls.crt | openssl x509 -noout -enddate` ? I assumed you used `openssl` to check this, but seeing the updated commit log and the comment above this one, I'm thinking you might have checked with `ls` instead?

praveenkumar replied:
I checked with openssl also and with openssl, timestamp is as expected but with ls it is not because certsyncer is handled by different controller which I mentioned in the commit log.
```
During testing of 4.17.0.ec.1 found out that
`secrets/aggregator-client/tls.crt` is synced before `configmaps/aggregator-client-ca/ca-bundle.crt` which is due to the certsyncer controller.
```

cfergeau replied:
I'd be worried if the openssl timestamps were out of order. The timestamps of the files on disk are less relevant, especially with the explanation you have now.

praveenkumar replied:
No it were not in the order, I can mention that in the commit log but since in our code we check the file so having that explanation would be also important.

cfergeau replied:
>  in our code we check the file 

We check it's modified/accessed/created time with `ls` / `stat` / ... ? I could only find use of `openssl x509 ...`

praveenkumar replied:
> We check it's modified/accessed/created time with `ls` / `stat` / ... ? I could only find use of `openssl x509 ...`

we check the cert expiry of the file using `openssl` command only and  this command run against  `/etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt` file,  during testing this file is not updated with rotated cert before `/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/aggregator-client/tls.crt` one hence this change.

cfergeau replied:
Can we tell somehow to the certsyncer controller it should flush the certs to disk? Adjusting our code to whatever it flushes to disk first feels like something which could change/break anytime, and need more changes, so it would be better if  we could tell it we want to work from a well-known state on disk?

praveenkumar replied:
>Can we tell somehow to the certsyncer controller it should flush the certs to disk? 

@cfergeau it is the controller which is not under our control and during the internal conversation around it looks like we can't do much around it.

>  Adjusting our code to whatever it flushes to disk first feels like something which could change/break anytime, and need more changes, so it would be better if we could tell it we want to work from a well-known state on disk?

Same feeling but looks like chicken and egg situation because it is updated to etcd first (which is source of truth) as part of configmap but since the api is not working until cert is rotated so we can't put code to pull it :(

cfergeau commented on pkg/crc/cluster/cert_renewal.go:
```diff
@@ -92,3 +92,11 @@ func waitForCertRenewal(sshRunner *ssh.Runner, cert string) func() error {
 	}
 }
+
+func WaitForKubeAPIServerCertRenewal(ctx context.Context, sshRunner *ssh.Runner, expired bool) error {
```
I'd add a `apiserver` boolean to `ApproveCSRAndWaitForCertsRenewal` and keep all the cert renewal-related code in this function.

praveenkumar replied:
sure this can be become helper function in that case.

praveenkumar replied:
Done

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -557,6 +557,10 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
+		return nil, errors.Wrap(err, "Failed to renew Kube API server cert: please check if a newer CRC release is available")
+	}
+
 	if err := cluster.WaitForAPIServer(ctx, ocConfig); err != nil {
```
Isn't this `WaitForAPIServer` call also retrying multiple times in case the `oc` calls fail because of an expired certificate?

praveenkumar replied:
yes it does but only for 2 mins and generation of cert happen in around ~7-8 mins.

cfergeau commented:
I assume these changes also work on a 4.16 cluster with no significant impact on startup time/.. when cert renewal is involved?

praveenkumar commented:
> I assume these changes also work on a 4.16 cluster with no significant impact on startup time/.. when cert renewal is involved?

Yes it works and it is not going to impact startup time in case cert is not expired and even it is expired it should add that ~8-10 mins which we already expect.

praveenkumar committed:
start: Remove workaround to delete apiserver pod in case of aggregator-client-ca regenerated

This workaround was required during OCP-4.3 timeframe and that bug is
also fixed so better to remove that workaround also.

- https://bugzilla.redhat.com/show_bug.cgi?id=1795163

cfergeau commented on pkg/crc/cluster/cert_renewal.go:
```diff
@@ -77,6 +77,10 @@ func ApproveCSRAndWaitForCertsRenewal(ctx context.Context, sshRunner *ssh.Runner
 	}
+	if aggregratorClient {
+		logging.Info("Kube API server certificate has expired, waiting for automatic renewal... [will take up to 8 minutes]")
+		return crcerrors.Retry(ctx, 8*time.Minute, waitForCertRenewal(sshRunner, AggregatorClientCert), time.Second*5)
```
Do we need 8 minutes just for this cert ?

praveenkumar replied:
@cfergeau yes it takes around ~7-8 mins

cfergeau replied:
7/8 minutes on top of the waits for the kubelet serving certificate and the kubelet client certificate, each of which can take up to 5 minutes? That feels like a lot

praveenkumar replied:
@cfergeau no I think during that time even those cert also rotating so it is mostly overall time for any cert expire

cfergeau reviewed (approved):
Not great that we cannot force the certoperator to sync its state to disk, and have to hope its external behaviour won't change too often :-/
